cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataSubscriberService.java

```diff
@@ -286,4 +296,48 @@ public void processMessage(MetadataMessage message) {
       }
     }
   }
+
+  private class MetadataOperationProcessor implements MetadataMessageProcessor {
+
+    @Override
+    public void processMessage(MetadataMessage message)  {
+      MetadataOperation operation = message.getPayload(GSON, MetadataOperation.class);
+      Metadata metadata = operation.getMetadata();
+      MetadataEntity entity = operation.getEntity();
+
+      // TODO: Authorize that the operation is allowed. Currently MetadataMessage does not carry user info
+      switch (operation.getType()) {
+        case PUT: {
+          LOG.trace("Received PUT for entity {}: {}", entity, metadata);
+          try {
+            if (metadata.getProperties() != null && !metadata.getProperties().isEmpty()) {
+              metadataAdmin.addProperties(entity, metadata.getProperties());
+            }
+            if (metadata.getTags() != null && !metadata.getTags().isEmpty()) {
+              Set<String> toAdd = metadata.getTags();
+              metadataAdmin.addTags(entity, toAdd.toArray(new String[toAdd.size()]));
```

Comment:
Can we have a `addTags` that takes a `Set` so that we don't need to copy the tags to an array?

Reply:
done. Changed them all to accept a set, and only kept the varargs version for the dataset tests